component/prometheus/operator/podmonitors/operator.go

```diff
@@ -0,0 +1,133 @@
+package podmonitors
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log/level"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/component/prometheus/scrape"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name: "prometheus.operator.podmonitors",
+		Args: Arguments{},
+
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args)
+		},
+	})
+}
+
+type Component struct {
+	mut     sync.Mutex
+	config  *Arguments
+	manager *CRDManager
+
+	onUpdate  chan struct{}
+	opts      component.Options
+	healthMut sync.RWMutex
+	health    component.Health
+}
+
+func (c *Component) CurrentHealth() component.Health {
+	c.healthMut.RLock()
+	defer c.healthMut.RUnlock()
+	return c.health
+}
+
+func New(o component.Options, args component.Arguments) (*Component, error) {
+	c := &Component{
+		opts:     o,
+		onUpdate: make(chan struct{}, 1),
+	}
+	return c, c.Update(args)
+}
+
+// Run implements component.Component.
+func (c *Component) Run(ctx context.Context) error {
+	// innerCtx gets passed to things we create, so we can restart everything anytime we get an update.
+	// Ideally, this component has very little dynamic config, and won't have frequent updates.
+	var innerCtx context.Context
+	// cancel is the func we use to trigger a stop to all downstream processors we create
+	var cancel func()
+	defer func() {
+		if cancel != nil {
+			cancel()
+		}
+	}()
+
+	c.reportHealthy()
+	errChan := make(chan error)
+	for {
+		select {
+		case <-ctx.Done():
+			if cancel != nil {
+				cancel()
+			}
+			return nil
+		case err := <-errChan:
+			c.reportUnhealthy(err)
+		case <-c.onUpdate:
+			if cancel != nil {
+				cancel()
+			}
+			innerCtx, cancel = context.WithCancel(ctx)
+			c.mut.Lock()
+			componentCfg := c.config
+			manager := NewCRDManager(c.opts, c.opts.Logger, componentCfg)
+			c.manager = manager
+			c.mut.Unlock()
+			go func() {
+				if err := manager.Run(innerCtx); err != nil {
+					level.Error(c.opts.Logger).Log("msg", "error running crd manager", "err", err)
+					errChan <- err
+				}
+			}()
+		}
+	}
+}
+
+// Update implements component.Component.
+func (c *Component) Update(args component.Arguments) error {
```

Comment:
In the future, it'd be nice to have Update initialize as much as it can so it can return errors early back to the caller. 

Can we add a TODO here for that?

Reply:
do you mean doing the CRDManager initialization in this function? do we even need the channel then? (you mention returning errors back early, here we always return `nil`, and I don't see any errors being returned outside of newly spawned goroutine, so it might not be necessary?)

Reply:
Yes, CRDManager.Run currently both initializes _and_ runs things, but initializing things could fail:

1. Initializing the rest client, scheme, and label selector could fail because of invalid config
2. Creating a cache for the informers might fail
3. Creating informers before running them could fail 

It would be nice to decouple the initialization logic from the running logic so we can report those errors early in the Update call before the component is running. 

(I don't think this should be done in this PR, definitely something for later)